core/src/block.rs

```diff
@@ -12,594 +11,798 @@
-            consensus_estimation: DEFAULT_CONSENSUS_ESTIMATION_MS,
-            height: self.height,
-            view_change_index: self.view_change_index,
-            previous_block_hash: self.previous_block_hash,
-            transactions_hash: None,
-            rejected_transactions_hash: None,
-            committed_with_topology: self.committed_with_topology.sorted_peers,
-        };
+    /// First stage in the life-cycle of a [`Block`].
+    /// In the beginning the block is assumed to be verified and to contain only accepted transactions.
+    /// Additionally the block must retain events emitted during the execution of on-chain logic during
+    /// the previous round, which might then be processed by the trigger system.
+    #[derive(Debug, Clone)]
+    pub struct Pending {
+        /// Unix timestamp
+        timestamp_ms: u128,
+        /// Collection of transactions which have been accepted.
+        /// Transaction will be validated when block is chained.
+        transactions: Vec<VersionedAcceptedTransaction>,
+        /// The topology at the time of block commit.
+        commit_topology: Topology,
+        /// Event recommendations for use in triggers and off-chain work
+        event_recommendations: Vec<Event>,
+    }
 
-        let mut txs = Vec::new();
-        let mut rejected = Vec::new();
+    impl BlockBuilder<Pending> {
+        /// Create [`Self`]
+        ///
+        /// # Panics
+        ///
+        /// if the given list of transaction is empty
+        #[inline]
+        pub fn new(
+            transactions: Vec<VersionedAcceptedTransaction>,
+            commit_topology: Topology,
+            event_recommendations: Vec<Event>,
+        ) -> Self {
+            assert!(!transactions.is_empty(), "Empty block created");
+
+            Self(Pending {
+                timestamp_ms: iroha_data_model::current_time().as_millis(),
+                transactions,
+                commit_topology,
+                event_recommendations,
+            })
+        }
 
-        for tx in self.transactions {
-            match self
-                .transaction_validator
-                .validate(tx.into_v1(), header.is_genesis(), &self.wsv)
-            {
-                Ok(tx) => txs.push(tx),
-                Err(tx) => {
-                    iroha_logger::warn!(
-                        reason = %tx.as_v1().rejection_reason,
-                        caused_by = ?tx.as_v1().rejection_reason.source(),
-                        "Transaction validation failed",
-                    );
-                    rejected.push(tx)
-                }
+        fn make_header(
+            timestamp_ms: u128,
+            height: u64,
+            previous_block_hash: Option<HashOf<BlockPayload>>,
+            view_change_index: u64,
+            transactions: &[VersionedValidTransaction],
+            rejected_transactions: &[VersionedRejectedTransaction],
+            commit_topology: Topology,
+        ) -> BlockHeader {
```

Comment:
`BlockHeaderBuilder` would indicate that there are some optional fields but, except for the genesis block, there are none

Reply:
I tried, but I still think this is useless. I create `RawHeader` only to have it destructured into `BlockHeader` in the next step. Because `make_header` takes arguments and assembles them into `BlockHeader`

Reply:
Ok, I don't insist. That won't be used in many places so the current version is ok for me